core/offchain/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Substrate offchain workers.
+//!
+//! The offchain workers is a special function of the runtime that
+//! gets executed after block is imported. During execution
+//! it's able to asynchronously submit extrinsics that will either
+//! be propagated to other nodes (transactions) or will be
+//! added to the next block produced by the node as inherents.
+//!
+//! Offchain workers can be used for computation-heavy tasks
+//! that are not feasible for execution during regular block processing.
+//! It can either be tasks that no consensus is required for,
+//! or some form of consensus over the data can be built on-chain
+//! for instance via:
+//! 1. Challenge period for incorrect computations
+//! 2. Majority voting for results
+//! 3. etc
+
+#![warn(missing_docs)]
+
+use std::{
+	marker::PhantomData,
+	sync::Arc,
+};
+
+use client::runtime_api::ApiExt;
+use inherents::pool::InherentsPool;
+use log::{debug, warn};
+use primitives::ExecutionContext;
+use runtime_primitives::{
+	generic::BlockId,
+	traits::{self, ProvideRuntimeApi},
+};
+use tokio::runtime::TaskExecutor;
+use transaction_pool::txpool::{Pool, ChainApi};
+
+mod api;
+
+pub use offchain_primitives::OffchainWorkerApi;
+
+/// An offchain workers manager.
+#[derive(Debug)]
+pub struct OffchainWorkers<C, Block: traits::Block> {
+	client: Arc<C>,
+	inherents_pool: Arc<InherentsPool<<Block as traits::Block>::Extrinsic>>,
+	executor: TaskExecutor,
+	_block: PhantomData<Block>,
+}
+
+impl<C, Block: traits::Block> OffchainWorkers<C, Block> {
+	/// Creates new `OffchainWorkers`.
+	pub fn new(
+		client: Arc<C>,
+		inherents_pool: Arc<InherentsPool<<Block as traits::Block>::Extrinsic>>,
+		executor: TaskExecutor,
+	) -> Self {
+		Self {
+			client,
+			inherents_pool,
+			executor,
+			_block: PhantomData,
+		}
+	}
+}
+
+impl<C, Block> OffchainWorkers<C, Block> where
+	Block: traits::Block,
+	C: ProvideRuntimeApi,
+	C::Api: OffchainWorkerApi<Block>,
+{
+	/// Start the offchain workers after given block.
+	pub fn on_block_imported<A>(
+		&self,
+		number: &<Block::Header as traits::Header>::Number,
+		pool: &Arc<Pool<A>>,
+	) where
+		A: ChainApi<Block=Block> + 'static,
+	{
+		let runtime = self.client.runtime_api();
+		let at = BlockId::number(*number);
```

Comment:
One last question: why block is identified by number (i.e. not the hash)? What if not best block is imported? Then the offchain worker could be executed with state of some other block (and probably it could be executed several times if there are several competing forks of the height `number`). Is this intended?

Reply:
Yes, I think that's expected. Since the offchain worker produces inherents for next block they might look different if an alternative fork is imported.
We might implement a way to invalidate previously generated inherents/extrinsics though, however 
offchain workers are also supposed to have a local storage in the future, so two "competing" tasks could have some sort of communication between runs i.e. the second run for block at height `X` will be able to tell it's on a fork that now became canon and for instance submit extrinsics that will replace the previous ones.

Numeric block number will also later be used by the offchain worker code to calculate lifetime of produced extrinsics (i.e. you can say I submit this inherent after performing a super-long-running computation, but it's only valid before block `bn + 3`), or you can also use that to execute some computation every X blocks.